parsers/CO.py

```diff
@@ -40,6 +57,163 @@ def fetch_consumption(
+        df_consumption = df_consumption.iloc[:, 2:26]
+
+        # Make sure values are positive and in MW
+        df_consumption = df_consumption[df_consumption >= 0] / 1000
+
+        hour = 0
+
+        datapoints = list()
+
+        for column in df_consumption.columns:
+
+            datapoint_datetime = datetime.combine(co_datetime.date(), time(hour))
+
+            datapoint = {
+                "zoneKey": zone_key,
+                "datetime": arrow.get(datapoint_datetime).replace(tzinfo=TZ).datetime,
+                "consumption": df_consumption[column].values[0],
+                "source": "xm.com.co",
+            }
+
+            datapoints.append(datapoint)
+
+            hour += 1
+
+        return datapoints
+
+
+@refetch_frequency(timedelta(days=1))
+def fetch_production(
+    zone_key: str = "CO",
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> dict:
+
+    # Convert datetime to local time
+    if target_datetime:
+        co_datetime = arrow.get(target_datetime).to(TZ)
+    else:
+        co_datetime = arrow.get(datetime.now() + timedelta(days=-3)).to(TZ)
+
+    objetoAPI = pydataxm.ReadDB()
+
+    # API request list of power plants with ID (column 1) and type (column 7)
+    df_recursos = objetoAPI.request_data(
+        "ListadoRecursos", "Sistema", co_datetime.date(), co_datetime.date()
+    )
+
+    # API request generation per power plant
+    df_generation = objetoAPI.request_data(
+        "Gene", "Recurso", co_datetime.date(), co_datetime.date()
+    )
+
+    if df_generation.empty:
+        return []
+
+    else:
+
+        # Add column with XM type
+        df_generation["XM Type"] = df_generation["Values_code"].apply(
+            lambda x: df_recursos.loc[
+                df_recursos["Values_Code"] == x, "Values_EnerSource"
+            ].values[0]
+        )
+
+        # Add column with EMaps type from types dict
+        df_generation["EM Type"] = df_generation["XM Type"].apply(
+            lambda x: TYPES_XM_EM.get(x) if x in TYPES_XM_EM else "unknown"
+        )
+
+        # Get the generation values per hour and per EM type of production
+        df_generation_type = df_generation.groupby(["EM Type"]).sum()
+
+        # Make sure values are positive expressed in MW
+        df_generation_type = df_generation_type[df_generation_type >= 0] / 1000
+
+        hour = 0
+
+        datapoints = list()
+
+        for column in df_generation_type.columns:
+
+            datapoint_datetime = datetime.combine(co_datetime.date(), time(hour))
+
+            datapoint = {
+                "zoneKey": zone_key,
+                "datetime": arrow.get(datapoint_datetime).replace(tzinfo=TZ).datetime,
+                "production": df_generation_type[column].to_dict(),
+                "source": "xm.com.co",
+            }
+
+            datapoints.append(datapoint)
+
+            hour += 1
+
+        return datapoints
```

Comment:
I think you should consider using the new classes Robin has made for this purpose, should make the parser more future proof.